hazelcast-jet-core/src/main/java/com/hazelcast/jet/windowing/SlidingWindowP.java

```diff
@@ -100,34 +102,38 @@ protected boolean tryProcessPunc0(@Nonnull Punctuation punc) {
     }
 
     private Traverser<Object> slidingWindowTraverser(Punctuation punc) {
-        Traverser<Object> traverser = Traversers.traverseStream(
-                range(nextFrameSeqToEmit, nextFrameSeqToEmit = wDef.higherFrameSeq(punc.seq()), wDef.frameLength())
-                        .mapToObj(frameSeq ->
-                                computeWindow(frameSeq, seqToKeyToFrame.remove(frameSeq - wDef.windowLength()))
-                                        .entrySet().stream()
-                                        .map(e -> new Frame<>(frameSeq, e.getKey(), finishF.apply(e.getValue()))))
-                        .flatMap(identity()));
-
-        traverser = traverser.append(punc);
-        return traverser;
+        long rangeStart = nextFrameSeqToEmit;
+        nextFrameSeqToEmit = wDef.higherFrameSeq(punc.seq());
+        return Traversers.traverseStream(range(rangeStart, nextFrameSeqToEmit, wDef.frameLength()).boxed())
+                .flatMap(frameSeq -> Traversers.traverseIterable(computeWindow(frameSeq).entrySet())
+                        .map(e -> (Object) new Frame<>(frameSeq, e.getKey(), finishF.apply(e.getValue())))
+                        .onFirstNull(() -> completeWindow(frameSeq)))
+                .append(punc);
     }
 
-    private Map<K, F> computeWindow(long frameSeq, Map<K, F> evictedFrame) {
+    private void completeWindow(long frameSeq) {
+        Map<K, F> evictedFrame = seqToKeyToFrame.remove(frameSeq - wDef.windowLength() + wDef.frameLength());
         if (deductF != null) {
-            // add leading-edge frame
-            patchSlidingWindow(combineF, seqToKeyToFrame.get(frameSeq));
             // deduct trailing-edge frame
             patchSlidingWindow(deductF, evictedFrame);
-            return slidingWindow;
         }
-        // without deductF we have to recompute the window from scratch
-        Map<K, F> window = new HashMap<>();
-        Map<Long, Map<K, F>> frames = seqToKeyToFrame.subMap(frameSeq - wDef.windowLength(), false, frameSeq, true);
-        for (Map<K, F> keyToFrame : frames.values()) {
-            keyToFrame.forEach((key, currAcc) ->
-                    window.compute(key, (x, acc) -> combineF.apply(acc != null ? acc : createF.get(), currAcc)));
+    }
+
+    private Map<K, F> computeWindow(long frameSeq) {
+        if (deductF != null) {
+            // add leading-edge frame
+            patchSlidingWindow(combineF, seqToKeyToFrame.get(frameSeq));
+            return slidingWindow;
+        } else {
```

Comment:
It was intentional: at first glance you see, that there are two paths that can be taken, each with it's own `return`. However, I consider this a matter of style, which I don't consider that important. But I would add [`InnerAssignmentCheck`](http://checkstyle.sourceforge.net/apidocs/com/puppycrawl/tools/checkstyle/checks/coding/InnerAssignmentCheck.html) to checkstyle...